packages/samples/headless-react/src/pages/DependentFacetPage.tsx

```diff
@@ -36,116 +35,241 @@ const Examples = {
 
 type ExampleType = keyof typeof Examples;
 
-export class DependentFacetPage extends Component<
-  {},
-  {currentExample: ExampleType}
-> {
-  private engine: SearchEngine;
-  private readonly authorFacet: HeadlessFacet;
-  private readonly sourceFacet: HeadlessFacet;
-  private readonly objectTypeFacet: HeadlessFacet;
-  private readonly fileTypeFacet: HeadlessFacet;
+type FacetWithDependenciesControllerOptions = FacetOptions &
+  Required<Pick<FacetOptions, 'facetId'>>;
 
-  constructor(props: {}) {
-    super(props);
+interface FacetWithDependenciesProps {
+  engine: SearchEngine;
+  controllerOptions: FacetWithDependenciesControllerOptions;
+  conditions?: AnyFacetValuesCondition<FacetValueRequest>[];
+}
 
-    this.engine = buildSearchEngine({
-      configuration: getSampleSearchEngineConfiguration(),
-    });
+const FacetWithDependencies: FunctionComponent<FacetWithDependenciesProps> = ({
+  engine,
+  controllerOptions,
+  conditions,
+}) => {
+  // Initialize facet.
+  const controller = useMemo(
+    () => buildFacet(engine, {options: controllerOptions}),
+    [engine, controllerOptions]
+  );
+  const [state, setState] = useState(controller.state);
+  useEffect(
+    () => controller.subscribe(() => setState(controller.state)),
+    [controller]
+  );
 
-    this.authorFacet = buildFacet(this.engine, {options: {field: 'author'}});
-    this.sourceFacet = buildFacet(this.engine, {options: {field: 'source'}});
-    this.objectTypeFacet = buildFacet(this.engine, {
-      options: {field: 'objecttype'},
-    });
-    this.fileTypeFacet = buildFacet(this.engine, {
-      options: {field: 'filetype'},
+  // Initialize conditions.
+  useEffect(() => {
+    if (!conditions?.length) {
+      return;
+    }
+    const conditionsManager = buildFacetConditionsManager(engine, {
+      facetId: controllerOptions.facetId,
```

Comment:
When switching tabs, the `state` in this `useEffect` call always contains an old facet, therefore `state.facetId` cannot be used, which isn't very intuitive.

A solution for this may be to replace `useState`/`useMemo` of solution A with something else so that the `state` always contains the correct value, I haven't had time to investigate this yet. Solution B or C would inherently solve this problem.